Finetuning pipeline

WenkelF opened the pull request:
@DomInvivo as discussed a first draft for the Finetuning pipeline.

Two possible pipelines:
`expts/main_run_finetuning_v1.py` (probably to be removed):
- Start with pretrained model
- Remove redundant task heads
- Drop/add layers of model part to be finetuned
- Drawback: Difficult updating of hyperparameters
`expts/main_run_finetuning_v2.py`:
- Initialize new model with final finetuning architecture
- Overwrite parameters that are shared with pretrained model

Main TODOs:
- Use hydra to easy finetuning config
- Create unit test with dummy pretrained model

DomInvivo commented on expts/main_run_finetuning_v2.py:
```diff
@@ -0,0 +1,153 @@
+    # Changes for finetuning
+
+    # Modify config
+    cfg = modify_cfg_for_finetuning(cfg)
```
Look into hydra

DomInvivo commented on expts/main_run_finetuning_v2.py:
```diff
@@ -0,0 +1,153 @@
+    logger.info(metrics)
+
+    predictor = load_predictor(
+        cfg, model_class, model_kwargs, metrics, accelerator_type, datamodule.task_norms
```
I would copy the `model_kwargs` from the `pretrained_model`, but modify the configs with a similar logic to `modify_cfg_for_pretrained`

DomInvivo commented on graphium/finetuning/finetuning.py:
```diff
@@ -0,0 +1,56 @@
+
+        cfg_finetune = cfg['trainer']['finetuning']
+        self.depth = cfg_finetune['added_layers']['depth']
+        try:
+            self.depth += cfg_finetune['add_finetune_depth']
+        except:
+            pass
```
I guess this is in case `'add_finetune_depth'` is missing? You an do this instead, which puts a default value of 0 if the key is missing. But I would suggest changing the `__init__` altogether
```suggestion
        self.depth += cfg_finetune.get("add_finetune_depth", 0)
```

DomInvivo commented on graphium/finetuning/finetuning.py:
```diff
@@ -0,0 +1,56 @@
+class MolecularFinetuning(BaseFinetuning):
+    def __init__(
+        self,
+        cfg,
```
```suggestion
        added_depth: int,
        on_task: str,
        epoch_unfreeze_all: int,
        add_finetune_depth: int = 0,
        ...
```

DomInvivo commented on graphium/finetuning/finetuning.py:
```diff
@@ -0,0 +1,56 @@
+from pytorch_lightning.callbacks import BaseFinetuning
+
+
+class MolecularFinetuning(BaseFinetuning):
```
Nothing from the class suggests that it's for Molecules only. I would cal it something more generic, like `GraphFinetuning` or `FullGraphNetworkFinetuning` (if specific to the `FullGraphNetwork` class, which it seems to be since it's using `model.task_heads` somewhere)

DomInvivo commented on graphium/finetuning/finetuning.py:
```diff
@@ -0,0 +1,56 @@
+        super().__init__()
+
+        cfg_finetune = cfg['trainer']['finetuning']
+        self.depth = cfg_finetune['added_layers']['depth']
```
Add some checks in the `__init__` that `depth` and `add_fine_depth` and `on_task` are consistent with the model. Perhaps this would require to pass the model at the `__init__`, which is weird, but we want the code to crash BEFORE we start processing the data and training the model. We want a quick crash, instead of waiting 20min+ before it crashes for a wrong depth.

DomInvivo commented on graphium/finetuning/finetuning.py:
```diff
@@ -0,0 +1,56 @@
+    def __init__(
+        self,
+        cfg,
+        train_bn: bool = False
```
what is `train_bn`? Name seems ambiguous

DomInvivo replied:
Ahhh I see! `train_bn` as in "should we continue training the bach_norm". Makes sense.

DomInvivo commented on graphium/finetuning/utils.py:
```diff
@@ -0,0 +1,27 @@
+from copy import deepcopy
+
+
+def modify_cfg_for_finetuning(cfg):
```
This could be a function within `FeedForwardNN` and `FullGraphNetwork`

WenkelF replied:
I am not sure, we can remove the function and replace inside the networks. As discussed in that #411, it might be possible to have a similar function within load_architecture in `graphium/data/_loader.py` but depends on the PR.

DomInvivo commented on graphium/nn/architectures/global_architectures.py:
```diff
@@ -1278,6 +1301,95 @@ def set_max_num_nodes_edges_per_graph(self, max_nodes: Optional[int], max_edges:
 
         self.task_heads.set_max_num_nodes_edges_per_graph(max_nodes, max_edges)
 
+    def modify_architecture(self, cfg):
```
Pass the parameters here, not the configs.
```suggestion
    def modify_architecture(self, choice, task, add_layers, out_dim, ...):
```

Also, as discussed, `choice` is an ambiguous name

DomInvivo commented on graphium/nn/architectures/global_architectures.py:
```diff
@@ -1278,6 +1301,95 @@ def set_max_num_nodes_edges_per_graph(self, max_nodes: Optional[int], max_edges:
+            self.task_heads.task_heads_kwargs[task].update(
+                {
+                    'out_dim': cfg_finetune['added_layers']['out_dim'],
+                    'depth': self.task_heads.task_heads_kwargs[task]['depth'] + cfg_finetune['added_layers']['depth'],
```
Be careful if some values are optional.

DomInvivo reviewed (commented):
Here's my review. It mostly aligns with what we discussed last time

cwognum commented on expts/finetune_configs/config_toy_finetuning_v2.yaml:
```diff
@@ -0,0 +1,259 @@
+
+  model_checkpoint:
+    dirpath: finetuned_models/
+    filename: finetuned-toy-model-{epoch}
```
It's a placeholder that will be used to format the string. See [`ModelCheckpoint`'s `filename` parameter](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.callbacks.ModelCheckpoint.html). Similar to normal string formatting:

```python
s = "{epoch}_text"
s.format(epoch=10)
# Returns: '10_text'

```

cwognum commented on graphium/nn/architectures/global_architectures.py:
```diff
@@ -1278,6 +1301,95 @@ def set_max_num_nodes_edges_per_graph(self, max_nodes: Optional[int], max_edges:
+                self.pre_nn.layers = pretrained_model.pre_nn.layers
+            except:
+                logger.warning(f"Pretrained ({pretrained_model.pre_nn}) and/or finetune model ({self.pre_nn}) do not use a pre-NN.")
+                pass
```
Just a warning? This seems like a major configuration error and should throw an Exception? 

(FYI:  `pass` is not needed here)

WenkelF replied:
Similar as above, this was part of finetuning v1, which we are not using. I have not removed the function because a similar function was already implemented by someone else (although it was not used anywhere).

Do you recommend removing the part of the code?

cwognum commented on graphium/nn/architectures/global_architectures.py:
```diff
@@ -1278,6 +1301,95 @@ def set_max_num_nodes_edges_per_graph(self, max_nodes: Optional[int], max_edges:
+        for module in ['pre_nn', 'pre_nn_edges', 'gnn', 'graph_output_nn', 'task_heads']:
+
+            if module == cfg_finetune['module_from_pretrained']:
+                break
```
`break` (exit the loop entirely) or `continue` (skip a single loop iteration)?

WenkelF replied:
Yes, because we only want to fully overwrite modules BEFORE the module we finetune from. The finetuning module is overwritten partially and modules afterwards are not overwritten (they keep their random initialization).

cwognum replied:
Ah check! I understand now, thanks!

cwognum reviewed (commented):
Hi @WenkelF, sorry for the delay here!

Thank you for this first implementation! 

I left comments whenever something stood out to me, but I'm aware that this PR is still WIP. Sorry if I pointed out some things that you were already planning to change. 

It would be super helpful if you could document the main fine-tuning "flow". Furthermore, I would suggest to simplify the process by adding support for one feature at a time, instead of having half-implemented features. This will make understanding, debugging, testing and maintaining the code base a lot easier. 

Happy to help next week!

DomInvivo commented on expts/hydra-configs/dataset/lipophilicity.yaml:
```diff
@@ -0,0 +1,309 @@
+      residual_type: none
+
+  task_heads:
+    qm9:
+      task_level: graph
+      out_dim: 19
+      hidden_dims: 128
+      depth: 2
+      activation: relu
+      last_activation: none
+      dropout: *dropout
+      normalization: *normalization
+      last_normalization: "none"
+      residual_type: none
+    tox21:
+      task_level: graph
+      out_dim: 12
+      hidden_dims: 64
+      depth: 2
+      activation: relu
+      last_activation: sigmoid
+      dropout: *dropout
+      normalization: *normalization
+      last_normalization: "none"
+      residual_type: none
+    zinc:
+      task_level: graph
+      out_dim: 3
+      hidden_dims: 32
+      depth: 2
+      activation: relu
+      last_activation: none
```
This doesn't make sense. We should not have any architectural choice from the original pre-trained model in here. Only things that would change.

That way, we can take different pre-trained models that have different hparams/seed and fine-tune them all with the same file.

WenkelF replied:
I agree. The configurations are still structured in a way where we have access to both the full config of the pretrained model and the pretraining-related config. And the modify_cfg_for_finetuning function consolidates information to one config.

This will be fixed once we incorporate the new hydra config from #421. We will still need modify_cfg_for_finetuning as of now. Therefore, it could be good waiting for the final version.

DomInvivo commented on expts/main_run_finetuning.py:
```diff
@@ -0,0 +1,131 @@
+import wandb
+
+# Set up the working directory
+MAIN_DIR = dirname(dirname(abspath(graphium.__file__)))
+os.chdir(MAIN_DIR)
```
The issues you're mentioning for this file are very related to the issues from `main_run_multitask.py`, which is what this file is based on. Perhaps you could open a new PR where you change the path handling and make sure it works?

DomInvivo replied:
And perhaps take care of the wandb config at the same time?

WenkelF replied:
Thanks for the comments @cwognum. I created an issue related to the working directory and wandb.

DomInvivo commented on graphium/data/datamodule.py:
```diff
@@ -1959,8 +1961,11 @@ def get_data_hash(self):
         for task in self.task_specific_args.keys():
-            if "epoch_sampling_fraction" in args[task].keys():
-                args[task].pop("epoch_sampling_fraction")
+            try:
+                if "epoch_sampling_fraction" in args[task].keys():
+                    args[task].pop("epoch_sampling_fraction")
+            except:
+                pass
```
I don't even understand the point of having a `try` and an `if` there. `Dict.pop` works even if the key is not available.

But we need to make sure that `args[task]` is not used elsewhere, even outside the current function since dict are passed as pointers. We only want to remove `epoch_sampling` for the hash key. So I would suggest the following.

```suggestion
            args[task] = deepcopy(args[task])
            args[task].pop("epoch_sampling_fraction")
```

WenkelF replied:
Indeed, this is not an ideal fix. I wanted to investigate the issue a bit more.

We cannot use `args[task].pop("epoch_sampling_fraction")` because `args[task]` may be of class `DatasetProcessingParams` instead of `Dict`. In particular, `ADMETBenchmarkDataModule` makes use of `DatasetProcessingParams`.

This originates from changes in 4b82ba3bb049b7ac1376f93e9537bd1a5d3842e0. We were using a Dict in all configs I saw, although I see a comment `# To be replaced by a new class "DatasetParams"` everywhere it appears.

Will create issue and think about a fix.

DomInvivo commented on graphium/data/datamodule.py:
```diff
@@ -1959,8 +1961,11 @@ def get_data_hash(self):
+            try:
+                if "epoch_sampling_fraction" in args[task].keys():
+                    args[task].pop("epoch_sampling_fraction")
+            except:
+                pass
```
See suggestion below.

DomInvivo commented on graphium/finetuning/finetuning.py:
```diff
@@ -0,0 +1,72 @@
+
+
+class GraphFinetuning(BaseFinetuning):
+    def __init__(self, cfg, train_bn: bool = False):
```
Change to explicitly pass parameters. 
```suggestion
    def __init__(self, fine-tuning, architecture, module_from_pretrained, ....................., train_bn: bool = False):
```

WenkelF replied:
Addressed in 76e2ba6777935056030eefb9f82dfc0b53a3884d

DomInvivo commented on graphium/finetuning/finetuning.py:
```diff
@@ -0,0 +1,72 @@
+            modules = pl_module.model.task_heads.graph_output_nn
+        elif module == "task_heads":
+            modules = pl_module.model.task_heads.task_heads
+
```
`else: raise "Wrong module"`

WenkelF replied:
Addressed in 76e2ba6777935056030eefb9f82dfc0b53a3884d

DomInvivo commented on graphium/finetuning/finetuning.py:
```diff
@@ -0,0 +1,72 @@
+            self.unfreeze_and_add_param_group(modules=pl_module, optimizer=optimizer, train_bn=self.train_bn)
+
+    def freeze_complete_module(self, pl_module: pl.LightningModule, module: str):
+        if module == "pe_encoders":
+            modules = pl_module.model.encoder_manager
+        elif module == "pre_nn":
+            modules = pl_module.model.pre_nn
+        elif module == "pre_nn_edges":
+            modules = pl_module.model.pre_nn_edges
+        elif module == "gnn":
+            modules = pl_module.model.gnn
+        elif module == "graph_output_nn":
+            modules = pl_module.model.task_heads.graph_output_nn
+        elif module == "task_heads":
+            modules = pl_module.model.task_heads.task_heads
```
I would define all these in a dictionary `_module_map = {pe_encoders: pl_module.model.encoder_manager, ...}` directly in the `__init__`. That way, with inheritance, someone could modified the entries without copy-pasting all the logic.

`_module_map` can replace the `module_list` you already have.

But instead of a regular dict, using an `OrderedDict` would also allow you to say something like: "freeze everything before task_heads" in a very simple way.

DomInvivo commented on graphium/finetuning/finetuning_architecture.py:
It's a bad idea to have a `FullGraphFinetuningNetwork` that basically copy-pastes most of the functionality of `FullGraphNetwork`.

Either use inheritance, or implement the fine-tuning logic directly within `FullGraphNetwork`

DomInvivo commented on graphium/nn/architectures/global_architectures.py:
```diff
@@ -272,6 +273,28 @@ def _create_layers(self):
+        Remove the last layers of the model part.
+        """
+
+        assert depth >= 0
+        assert depth <= len(self.layers)
```
hmmm... We've used tons of `assert` everywhere in the code. Do you recommend doing that everytime:
```python
if depth < 0: raise "depth should be >= 0"
```

cwognum replied:
Yeah, I would. My understanding is that assert statements can be removed at compile time (to improve performance, which seems specifically relevant to Graphium), so any error that a downstream user can make and should be informed about should ideally be a raised and specific Exception. To have sanity checks during debugging / development, you can still use assert statements.

DomInvivo reviewed (commented):
Don't forget that the objective is to release a working, incomplete version first. Then refining it with more complex fine-tuning possibilities.

WenkelF commented:
@DomInvivo thanks for your comments. Here is also a quick overview of the updates

Updates:
- Moved finetuning-related architecture out of FullGraphMultitaskNetwork to FullGraphFinetuningNetwork under graphium/finetuning/finetuning_architecture
-- More flexibility to build finetuning head on top of pretrained model
-- For now, finetuning is specific to FullGraphMultitaskNetwork (as pretrained model), but we can finetune from gnn, graph_output_nn and task_heads now
-- Additionally, we can add a flexible FinetuningHead, inheriting from nn.Module and MupMixin. For example, it can be a FeedForwardNN, FeedForwardPyg, TaskHeads, or a custom network that can be added to FINETUNING_HEAD_DICT
-- FinetuningHead needs a bit more work to correctly support all possible scenarios (modules/levels to finetune from); if not used, we automatically fall back to the finetuning logic, where we can only drop & add depth to the module we finetune from
-- Finished functions for overwriting shared weights and partially freezing modules while training and extended them from only task_heads to also gnn and graph_output_nn
- Added preliminary unit test the discussed “minimal” finetuning pipeline (finetune from one of the task heads)
-- We currently test depths, in_dim, ou_dim of changed modules are correct and overwritten correctly
-- Test does not cover training/freezing yet
-- Planning to add tests for finetuning from other modules as well
- Configuration has been integrated with hydra but needs more work to be easier to use (@luis-mueller is helping with that)

Remarks:
- Hidden_dims can only be used as int together with depth at the moment
- Currently, everything has only been tested on cpu/gpu

WenkelF commented on expts/main_run_finetuning.py:
```diff
@@ -0,0 +1,131 @@
+
+    cfg = modify_cfg_for_finetuning(cfg)
+
+    run_name: str = "main"
+    add_date_time: bool = True
+
+    st = timeit.default_timer()
+
+    date_time_suffix = ""
+    if add_date_time:
+        date_time_suffix = datetime.now().strftime("%d.%m.%Y_%H.%M.%S")
```
I also added this part to the issue mentioned above.

WenkelF commented on graphium/nn/architectures/global_architectures.py:
```diff
@@ -1278,6 +1301,95 @@ def set_max_num_nodes_edges_per_graph(self, max_nodes: Optional[int], max_edges:
+            )
+
+        else:
+            raise NotImplementedError()
```
I had two different finetuning approaches (v1 & v2) implemented in the beginning. This is a relic of v1 and, which we will not be using. I removed the code by now.

cwognum commented on graphium/finetuning/utils.py:
```diff
@@ -0,0 +1,26 @@
+    module = cfg_finetune['module_from_pretrained']
+    task = cfg_finetune['task']
+    level = cfg_finetune['level']
+    task_head_from_pretrained = cfg_finetune['task_head_from_pretrained']
```
Fyi: You don't need the None here. That's the default value.

WenkelF commented on graphium/data/datamodule.py:
@cwognum, could the issue in `expts/test_multitask_datamodue.py` be caused by this recent change?

cwognum commented:
xref: https://github.com/WenkelF/graphium/pull/10

WenkelF commented:
This pull introduces the following:
- Finetuning pipeline
- Updates to hydra configuration (by @cwognum )

The finetuning pipeline is maintained separately from existing architectures under `graphium/finetuning/finetuning_architecture`. The FullGraphFinetuningNetwork comes with two submodules:
- `PretrainedModel`
- `FinetuningHead `(optional)

`PretrainedModel` can load pretrained models (e.g., FullGraphMultitaskNetwork) and allows basic finetuning operations (in particular from the task heads), including dropping or extending layers of the module to finetune from.

`FinetuningHead` is optional and allows to define fully customizable networks that are applied on top of PretrainedModel. If not specified, we fall back to basic finetuning explained above.

Training is handled by the `GraphFinetuning `callback in `graphium/finetuning/finetuning`.

All methods in [graphium/finetuning](https://github.com/WenkelF/graphium/tree/explore_finetuning/graphium/finetuning) are implemented such that they are not specific to a pretrained model or finetuning head. This is achieved by requiring the pretrained model to come with a module_map (see, e.g., `FullGraphMultitaskNetwork`) that facilitates setting pretrained weights and freezing the correct layers.

The new unit test `tests/test_finetuning` tests the pipeline end-to-end for a specific case. Together with the corresponding config, it might be a good starting point to understand the pipeling and what information is needed in the config.

The updates to hydra allow to easily switch between benchmarking and finetuning. Major changes are documented here https://github.com/WenkelF/graphium/pull/10

DomInvivo commented on graphium/trainer/predictor.py:
```diff
@@ -96,20 +95,46 @@ def __init__(
             eval_options[task].check_metrics_validity()
 
+        # Work-around to retain task level when model_kwargs are modified for FullGraphFinetuningNetwork
+        if "task_heads_kwargs" in model_kwargs.keys():
+            task_heads_kwargs = model_kwargs["task_heads_kwargs"]
+        elif "pretrained_model_kwargs" in model_kwargs.keys():
+            # This covers finetuning cases where we finetune from the task_heads
+            task_heads_kwargs = model_kwargs["pretrained_model_kwargs"]["task_heads_kwargs"]
+        else:
+            raise ValueError("incorrect model_kwargs")
```
I don't think this should be here. I think that, if you are using a pre-trained model, you should pass directly `model_kwargs["pretrained_model_kwargs"]` into `model_kwargs`.

WenkelF replied:
Removed in a3d471540bc3f69cc0af2346ab0e66141b65311b as explained below

DomInvivo commented on graphium/trainer/predictor.py:
```diff
@@ -96,20 +95,46 @@ def __init__(
         self._eval_options_dict = {
             self._get_task_key(
-                task_level=model_kwargs["task_heads_kwargs"][key]["task_level"], task=key
+                task_level=task_heads_kwargs[key]["task_level"],
+                task=key
+                # task_level=model_kwargs["task_heads_kwargs"][key]["task_level"], task=key
```
In general, the `PredictorModule` should be agnostic to the model passed. By having the `self._get_task_key` here, it forces a certain architecture in the config which is not very flexible. 

I see that this logic was introduced already in the code prior to this PR. If it requires too many changes, let's open a new issue.

WenkelF replied:
You are right, thanks for pointing out. 

We could achieve this by getting the task-specific information (which is only the task level as far as I know) from the datamodule.

Here is how this could be done:
a3d471540bc3f69cc0af2346ab0e66141b65311b

What do you think?

DomInvivo commented on graphium/trainer/predictor.py:
```diff
@@ -96,20 +95,46 @@ def __init__(
 
         self.model = self._model_options.model_class(**self._model_options.model_kwargs)
+
+        # Maintain module map to easily select modules
+        # We now need to define the module_map in pretrained_model in FinetuningNetwork
+        # self._module_map = OrderedDict(
+        #     pe_encoders=self.model.encoder_manager,
+        #     pre_nn=self.model.pre_nn,
+        #     pre_nn_edges=self.model.pre_nn_edges,
+        #     gnn=self.model.gnn,
+        #     graph_output_nn=self.model.task_heads.graph_output_nn,
+        #     task_heads=self.model.task_heads.task_heads,
+        # )
+
```
Delete this comment.

DomInvivo commented on graphium/trainer/predictor.py:
```diff
@@ -96,20 +95,46 @@ def __init__(
         loss_fun = {
             self._get_task_key(
-                task_level=model_kwargs["task_heads_kwargs"][key]["task_level"], task=key
+                task_level=task_heads_kwargs[key]["task_level"],
+                task=key
```
Again, don't like how the config structure is imposed. Perhaps `task_level` should simply be passed to the `PredictorModule` to keep flexibility

WenkelF replied:
Good idea, this is implemented in a3d471540bc3f69cc0af2346ab0e66141b65311b

DomInvivo commented on tests/dummy-pretrained-model-gpu.ckpt:
Why duplicating the model for CPU and GPU? Models should be agnostic to the training hardware, and to the fine-tuning hardware. You can train on CPU and fine-tune on GPU or IPU

DomInvivo replied:
Yes, you can remove the gpu model.

DomInvivo reviewed (changes requested):
Mostly looks good, great work on this major PR!
A few changes to make, and some comments.

DomInvivo commented:
@zhiyil-graphcore @s-maddrellmander We'll need your help here to fix the tests for IPU. And ideally, have a test that loads a CPU-trained model onto IPU for finetuning.

cwognum commented on graphium/cli/finetune.py:
Why not keep it general? There is no reason this should be limited to just ADMET and I would argue against having a CLI for each experiment. That will lead to a lot of code repetition!

DomInvivo replied:
There are lots of things hardcoded for admet. But I do agree, once we have more things to fine-tune on, we can remove the hardcoding and use the same file.

cwognum replied:
> There are lots of things hardcoded for admet. 

I don't think so? Only the filtering of the config, but that can easily be seen as a more general config design pattern.

cwognum replied:
I checked and it can be used with any of the training configs we had (the logic is the same as for `main_run_multitask.py`). Fine-tuning for now only works for ADMET, but as I commented above (and documented in the code), that can be easily extended. I will thus keep things named as is.

cwognum replied:
My bad! I was looking at the wrong file! Blame the jetlag! 😉 

I still think we can keep the filename the same, since the CLI method is already very specifically named and it is easy to imagine that this will be extended in the future. This also provides contributors with a clear location to add finetuning CLIs.

cwognum commented on expts/hydra-configs/tasks/admet.yaml:
Not sure I understand. The purpose is the same as any `tasks` config; Allows you to include these tasks in a training/fine-tuning loop!

cwognum replied:
If you're asking why we have a config to merge two subconfigs: That's mostly to remain backwards compatible. I had to split the config in two parts since we only need one part during finetuning, whereas we need both during training.

DomInvivo replied:
Ok. Do you think that using more explicit names such as `task_heads` and `data_and_metrics` would make sense?

cwognum commented on expts/hydra-configs/finetuning/admet.yaml:
I understand where you're coming from, but I personally don't think renaming is the solution. If you would create a config for each task, you end up with ~20 configs with lots of repetition. I would keep the name general, but make the variability more explicit with comments in the yaml or even by introducing new variables in the config.

DomInvivo replied:
Yeah, let's add some comments to show which parameters are usually overwritten in the other configs. Otherwise it's confusing

cwognum commented on env.yml:
```diff
@@ -28,7 +28,7 @@ dependencies:
 
   # ML packages
-  - cudatoolkit # works also with CPU-only system.
+  - cuda-version =11.8 # works also with CPU-only system.
```
Good catch! Copied this from @hadim his tutorial, but we likely don't need it!

cwognum commented on expts/hydra-configs/finetuning/finetuning.yaml:
```diff
@@ -0,0 +1,82 @@
+
+# Changes to be implemented
+
+# predictor:
```
This file no longer exists. Resolving the comment!

WenkelF commented:
@cwognum the bug in the finetuning training is fixed here febdf2dcf21885e0c39e02d7caf64833e4e969d1

I missed a deepcopy operation when defining the datahashes for the TDC datasets. We include the first 5 rows of the df when generating the hash and the bug reduced the datasets to those 5 rows (molecules) as well.

Make sure to remove the TDC datasets from datacache. You can use `datamodule._path_to_load_from_file("train)` to get "train_[data_hash]" and then remove all data hashes (e.g., "train_[data_hash]", "test_[data_hash]", etc. that end with the same sequence.

WenkelF commented:
@DomInvivo I added some final improvements:
1. Made `modify_cfg_for_finetuning` function fully agnostic to pretrained model
2. Dropped dummy-pretrained-model-gpu and added `map_location` as an argument to  `load_pretrained_models `function
3. Added the option to keep modules after finetuning module

(3.) makes it much easier to finetune from modules other than the task_heads without manually re-defining the downstream network. When finetuning from task_heads, it is not needed.

s-maddrellmander commented:
@WenkelF - try merging from the main branch, I've made a change to the IPU test CLI that should take into account the changes made in this PR. If that doesn't work let me know.
